File: src/cpp_pubsub_pkg/src/listener_custom.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/msg/location.hpp"
using std::placeholders::_1;

class Listener : public rclcpp::Node
{
  public:
    Listener()
    : Node("custom_listener")
    {
      subscription_ = this->create_subscription<custom_interfaces::msg::Location>(
      "custom_topic", 10, std::bind(&Listener::topic_callback, this, _1));
    }

  private:
    void topic_callback(const custom_interfaces::msg::Location & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "I heard: (%f, %f, %f)/%s", 
                      msg.location.x, msg.location.y,
                      msg.location.z, msg.name.c_str());
    }
    rclcpp::Subscription<custom_interfaces::msg::Location>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}

File: src/py_pubsub_pkg/py_pubsub_pkg/listener_custom.py
import rclpy
from rclpy.node import Node
from custom_interfaces.msg import Location


class CustomListener(Node):

    def __init__(self):
        super().__init__('custom_listener')
        self.subscription = self.create_subscription(
            Location,
            'custom_topic',
            self.topic_callback,
            10)
        self.subscription  # prevent unused variable warning

    def topic_callback(self, msg):
        self.get_logger().info(
            f'I heard: ({msg.location.x:.2f}, {msg.location.y:.2f}, {msg.location.z:.2f}) / {msg.name}'
        )


def main(args=None):
    rclpy.init(args=args)
    node = CustomListener()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()